src/QsCompiler/QirGeneration/Subtransformations/ExpressionKindTransformation.cs

```diff
@@ -386,6 +386,252 @@ IValue CopyAndUpdateUdt(TupleValue originalValue)
+
+                var createString = sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.StringCreate);
+                return sharedState.CurrentBuilder.Call(createString, zeroLengthString);
+            }
+
+            // Creates a new string with reference count 1 that needs to be queued for unreferencing
+            // and contains the concatenation of both values. Both both arguments are unreferenced.
+            Value DoAppend(Value? curr, Value next, bool unreferenceNext = true)
+            {
+                if (curr == null)
+                {
+                    return next;
+                }
+
+                // The runtime function StringConcatenate creates a new value with reference count 1.
+                var concatenate = sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.StringConcatenate);
+                var unreference = sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.StringUpdateReferenceCount);
+                var countChange = sharedState.Context.CreateConstant(-1);
+                var app = sharedState.CurrentBuilder.Call(concatenate, curr, next);
+                sharedState.CurrentBuilder.Call(unreference, curr, countChange);
+                if (unreferenceNext)
+                {
+                    sharedState.CurrentBuilder.Call(unreference, next, countChange);
+                }
+                return app;
+            }
+
+            // Creates a string value that needs to be queued for unreferencing.
+            Value ExpressionToString(IValue evaluated)
+            {
+                void UpdateStringRefCount(Value str, int change)
+                {
+                    var addReference = sharedState.GetOrCreateRuntimeFunction(RuntimeLibrary.StringUpdateReferenceCount);
+                    var countChange = sharedState.Context.CreateConstant(change);
+                    sharedState.CurrentBuilder.Call(addReference, str, countChange);
+                }
+
+                // Creates a string value that needs to be queued for unreferencing.
+                Value SimpleToString(string rtFuncName)
+                {
+                    var createString = sharedState.GetOrCreateRuntimeFunction(rtFuncName);
+                    return sharedState.CurrentBuilder.Call(createString, evaluated.Value);
+                }
+
+                Value TupleToString(TupleValue tuple)
+                {
+                    var str = CreateConstantString($"{tuple.TypeName}(");
+                    var tupleElements = tuple.GetTupleElements();
+                    Value? comma = null;
+
+                    for (var idx = 0; idx < tupleElements.Length; ++idx)
+                    {
+                        str = DoAppend(str, ExpressionToString(tupleElements[idx]));
+                        var isLast = idx == tupleElements.Length - 1;
+                        var next = isLast
+                            ? CreateConstantString(")")
+                            : comma ?? CreateConstantString(",");
```

Comment:
Minor:
(I don't know C#, I can be wrong)
I have an impression that `comma` is always `null` here.

Reply:
Eh, yes; clearly I've been very distracted writing this code...! Thanks for the proper review; it was sorely needed.